Update `CONTRIBUTING.md` (#2820)
spacedriveapp/spacedrive: CONTRIBUTING.md

 - Make sure to read the [guidelines](https://spacedrive.com/docs/developers/prerequisites/guidelines) to ensure that your code follows a similar style to ours.

 - After you finish making your changes and committed them to your branch, make sure to execute `pnpm autoformat` to fix any style inconsistency in your code.

### Landing Page

To run the **landing page**, run:

- `pnpm landing dev`

### Mobile App

To run the mobile app:

- Run `./scripts/setup.sh mobile`

#### Android

- Install Java JDK 17 for Android. Java 21 is [not compatible](https://github.com/react-native-async-storage/async-storage/issues/1057#issuecomment-1925963956).
- Install [Android Studio](https://developer.android.com/studio). This will set up most of the dependencies required to build the mobile app.
  - Make sure you have [NDK 26.1.10909125 and CMake](https://developer.android.com/studio/projects/install-ndk#default-version) installed in Android Studio.
- Run `pnpm mobile android` to build the core and begin debugging the app.

> [!TIP]
> To speed up compilation for android you may temporarily remove uneccecary architectures from the build by removing them from the following line:

> > https://github.com/spacedriveapp/spacedrive/blame/d180261ca5a93388486742e8f921e895e9ec26a4/apps/mobile/modules/sd-core/android/build.sh#L61

Reviewer: Hey, for some reason, this won't preview properly on GitHub. The other preview works great, though.

Once this is fixed, I'm good to let it merge.

Author: Seems like GitHub didn't like this line being linked from the blame view, switching it over fixed it (a2bb6e0fc68de8c9574c9faf1369a6aefdd431fc)

### Making Changes

#### Spacedrive Desktop Clients

This project uses [Cargo](https://doc.rust-lang.org/cargo/getting-started/installation.html) and [pnpm](https://pnpm.io/installation). Make sure you have them installed before proceeding.

To make changes locally, follow these steps:

1. Clone & enter the repository:
    ```
    git clone https://github.com/spacedriveapp/spacedrive && cd spacedrive
    ```

>     Alternately, if you already have the repo cloned from previously, pull the latest changes to the code with: `git pull`

Reviewer 2: ```suggestion
    Alternately, if you’ve already cloned the repo locally, pull the latest changes with: `git pull`
```

      > [!NOTE]
      > This [script](https://github.com/spacedriveapp/spacedrive/blob/main/scripts/setup.sh#L133) will check if Rust and pnpm are installed then proceed to install Clang, NASM, LLVM, libvips, Gstreamer's Plugins, FFmpeg, Perl, [Tauri essentials](https://tauri.app/v1/guides/getting-started/prerequisites/#setting-up-linux) and any other required dependencies for Spacedrive to build via your system's respective package manager.
  - For Windows users, run: `.\scripts\setup.ps1` via PowerShell.
      > [!NOTE]
      > This [script](https://github.com/spacedriveapp/spacedrive/blob/main/scripts/setup.ps1#L81) will install pnpm, LLVM, FFmpeg, C++ build tools, NASM, Rust + Cargo, Rust tools, Edge Webview 2, Strawberry Perl, [Tauri essentials](https://tauri.app/v1/guides/getting-started/prerequisites/#setting-up-windows) and any other required dependencies for Spacedrive to build.
3. Install NodeJS dependencies: `pnpm i`
4. Prepare the build: `pnpm prep`
    > [!NOTE]
    > This will run all necessary codegen and build required dependencies

> [!TIP]
> Linux & macOS users can download a bundle of sample files for testing via `pnpm test-data` (requires `curl` & `tar`)
>

> >  The test files will be located in a directory called `test-data` in the root of the spacedrive repo.

Reviewer 2: ```suggestion
>  The test files will be located in a directory called `test-data` in the root of the Spacedrive repo.
```

> To run the **Tauri desktop** app, run:

Reviewer 2: ```suggestion
To run the **desktop** app, run:
```

Author: I'm leaning towards including Tauri here since it's part of the command to build some mental connections for users.

Alternatively, perhaps we should just add another script to pnpm so we could run `pnpm desktop dev`, which might make more sense?

```
pnpm tauri dev
```
> [!NOTE]
> The Tauri desktop app always runs it's own instance of the backend and will not connect to a seperately initiated `sd-server` instance.

To run the **backend server**, run:
```
cargo run -p sd-server
```

> > [!TIP]
> > If necessary, the [webview devtools](https://tauri.app/v1/guides/debugging/application/#webview-console) can be opened by pressing `Ctrl + Shift + I` (Linux and Windows) or `Command + Option + I` (macOS) in the desktop app.
> >
> > Also, the react-devtools can be launched using `pnpm dlx react-devtools`.

Reviewer 2: ```suggestion
> [!TIP]
> If necessary, [DevTools](https://tauri.app/v1/guides/debugging/application/#webview-console) for the WebView can be opened by pressing <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>I</kbd> (Linux and Windows) or <kbd>Command</kbd>+<kbd>Option</kbd>+<kbd>I</kbd> (macOS) in the desktop app.
>
> Also, React DevTools can be launched using `pnpx react-devtools`.
```

  However, it must be executed before starting the desktop app for it to connect.

To run the **web** app (requires the backend to be running), run:
```
pnpm web dev
```
> [!TIP]
> You can also quickly launch the web interface together with the backend with:
> ```
> pnpm dev:web
> ```

To run the **e2e tests** for the web app:
```
pnpm web test:e2e
```
If you are developing a new test, you can execute Cypress in interactive mode with:
```
pnpm web test:interactive
```

##### Troubleshooting

- If you encounter any issues, ensure that you are using the following versions of Rust, Node.js and pnpm:
  | tool    | version |
  | ----    | ------- |
  | Rust    | [`1.81`](rust-toolchain.toml) |
  | Node.js | [`18.18`](.nvmrc) |
  | pnpm    | `9.4.0` |

  [`rustup`](https://rustup.rs/) & [`nvm`](https://github.com/nvm-sh/nvm) should both pick up on the appropriate versions of the Rust Toolchain & Node respectively from the project automatically. 

>  - After cleaning out your build artifacts using `pnpm clean`, it is necessary to re-run the `./setup/setup.sh` script.

Reviewer 2: It’s necessary to rerun `prep`, but not the setup script

```suggestion
 - After cleaning out your build artifacts using `pnpm clean`, it's necessary to re-run `pnpm prep`.
```
